SDALib-CPP/SDALibTests/SDADriverTest.cpp

```diff
@@ -16,87 +16,192 @@ void DriverSide()
-	ASSERT_EQ(server.SendData("STOP", 4), IPCLIB_SUCCEED);
-	server.AwaitData(buffer, TEST_BUFFER_SIZE);
-	ASSERT_TRUE(buffer[0] == 'O' && buffer[1] == 'K' && buffer[2] == '\0');
+	p_server.ReceiveDataAsync();
+	ASSERT_EQ(p_server.SendData("STOP", 4), IPCLIB_SUCCEED);
+	ASSERT_DURATION_LE(1, p_server.AwaitData(p_buffer, TEST_BUFFER_SIZE));
+	ASSERT_TRUE(p_buffer[0] == 'O' && p_buffer[1] == 'K' && p_buffer[2] == '\0');
 	std::this_thread::sleep_for(std::chrono::milliseconds(10)); // wait to disconnect client
-	server.CloseServer();
-}
\ No newline at end of file
+	p_server.CloseServer();
+}
+
+/// @brief				 The server side of the tests
+/// @param  p_server	 The server socket that has a connection to the driver
+/// @param  p_situations The amount of situations the driver has to handle
+void ServerSide(ServerSocket& p_server, int p_situations)
+{
+	ASSERT_DURATION_LE(3, while (!p_server.Connected()) {});
+	char buffer[TEST_BUFFER_SIZE];
+
+	TestActive(p_server, buffer);
+
+	TestOrder(p_server, buffer);
+
+	TestTestAmount(p_server, buffer);
+
+	TestSituation(p_server, buffer); // 1 test situation
+
+	for(int i = 0; i < p_situations; i++)
+	{
+		TestSituation(p_server, buffer);
+	}
+
+	TestStopConnection(p_server, buffer);
+}
+
+/// @brief				 Tests the behaviour of a driver when it does not have to wait on a connection of the server
+/// @param  p_situations The amount of different situations the driver has to handle
+void NoWaitTest(int p_situations)
+{
+	// connect server and black box
+	ServerSocket server;
+	ASSERT_EQ(server.Initialize(), IPCLIB_SUCCEED);
+	server.ConnectAsync();
+	std::thread t = std::thread(DriverSide);
+	t.detach();
+
+	ServerSide(server, p_situations);
+}
+
+/// @brief Tests a full session with a server
+TEST(DriverTests, NoWaitNoSituations)
+{
+	NoWaitTest(0);
+}
+
+/// @brief Tests a full session with a server
+TEST(DriverTests, NoWaitOneSituation)
+{
+	NoWaitTest(1);
+}
+
+/// @brief Tests a full session with a server
+TEST(DriverTests, NoWaitMultipleSituations)
+{
+	NoWaitTest(4);
+}
+
+/// @brief				 Tests the behaviour of a driver if it can't connect to a server
+/// @param  p_waitAmount The time the driver has to wait
+/// @param  p_situations The amount of different situations the driver has to handle
+void WaitTest(int p_waitAmount, int p_situations)
+{
+	std::thread t = std::thread(DriverSide);
+	t.detach();
+
+	std::this_thread::sleep_for(std::chrono::seconds(p_waitAmount));
+
+	ServerSocket server;
+	ASSERT_EQ(server.Initialize(), IPCLIB_SUCCEED);
+	server.ConnectAsync();
+
+	ServerSide(server, p_situations);
+}
+
+TEST(DriverTests, WaitNoSituations)
+{
+	WaitTest(3, 0);
+}
+
+TEST(DriverTests, WaitOneSituation)
+{
+	WaitTest(3, 1);
+}
+
+TEST(DriverTests, WaitMultipleSituations)
+{
+	WaitTest(3, 4);
+}
+
+
+TEST(DriverTests, DoubleWaitNoSituations)
+{
+	WaitTest(5, 0);
+}
+
+TEST(DriverTests, DoubleWaitOneSituation)
+{
+	WaitTest(5, 1);
+}
+
+TEST(DriverTests, DoubleWaitMultipleSituations)
+{
+	WaitTest(5, 4);
```

Comment:
Add the BEGIN_TEST_COMBINATORIAL() and END_TEST_COMBINATORIAL2() to the Utils.h and use that